src/main/java/io/numaproj/numaflow/sinker/DatumIteratorImpl.java

```diff
@@ -0,0 +1,54 @@
+package io.numaproj.numaflow.sinker;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+
+/**
+ * An implementation of {@link DatumIterator} that uses a blocking queue to store the elements.
+ */
+@Slf4j
+class DatumIteratorImpl implements DatumIterator {
+    private final BlockingQueue<Datum> blockingQueue = new LinkedBlockingDeque<>();
+    private Datum nextElement;
+    private boolean isClosed = false;
+
+    /*
+     * Returns true if there is at least one more element in the iterator, or false otherwise.
+     * This method blocks until an element becomes available in the blocking queue.
+     * When EOF is received, this method will return false and the iterator will be closed.
+     */
+    @Override
+    public boolean hasNext() {
+        if (nextElement != null) {
+            return true;
+        }
+        try {
+            nextElement = blockingQueue.take(); // block until an element becomes available
```

Comment:
This approach of taking an element out of the queue within `hasNext()` doesn't feel right. Calling it multiple times will return different results, but it shouldn't. (Please add this case to unit tests after we fix this.)

`hasNext()` is not supposed to modify the underlying data structure. Consider using `blockingQueue.peek()` to implement it. Please also see if we can simplify by removing the `nextElement` and the `isClosed` flag.

Reply:
Had an offline discussion with Karen and we decided to just go with `next()` and remove `hasNext()` for two reasons
- There is a tight dependency between `hasNext()` and `next()`, where `next()` should be called immediately after `hasNext()`. If we were to call `hasNext()` without calling `next()`, the message would be lost.
- To ensure thread safety, it is currently not possible to share the DatumStream among multiple threads with both `hasNext() `and `next()`.